### ISSUE TITLE
DistanceMap least cost routing does not take ScramDrive 150 fuel into account

### ISSUE THREAD

**Issue by EndrosG** [endless-sky/endless-sky#2435]
as the title says, the DistanceMap class does not take into account, whether the player has a regular Hyperdrive or a Scram Drive when deciding between following links or jumping to neighbors.

As I first created a draft in GIMP because I didnt find a proper candidate on the map in the first place, I now attach two images. The draft and the place which I later found on the real map.
And of course a savegame.
[Testing Tester~Scram Drive.txt](https://github.com/endless-sky/endless-sky/files/944906/Testing.Tester.Scram.Drive.txt)
Savegame contains:
* Hawk with 1000x hull matrix so you'll survive jumping through Korath space
* equipped with Scram Drive and Jump Drive, fuel tank modded to 500
* located at Shangri-La, which has a shipyard + outfitter (which sells regular Hyperdrives, in case you want to test something)

![endless_sky_v0_9_6__20170420__pathfinding](https://cloud.githubusercontent.com/assets/26883037/25243975/3829d0ba-25ff-11e7-97c1-d74c51adfc8d.png)
![endless_sky_v0_9_6__20170420__pathfinding_draft](https://cloud.githubusercontent.com/assets/26883037/25243978/39df2266-25ff-11e7-8843-3705132340ef.png)

The 3 jumps via links cost 3x 150 = 450 in total with Scram Drives.
The 2 direct jumps using the jump drive cost 2x 200 = 400 in total.

	// Find the route with lowest fuel use. If multiple routes use the same fuel,
	// choose the one with the fewest jumps (i.e. using jump drive rather than
	// hyperdrive). If multiple routes have the same fuel and the same number of
	// jumps, break the tie by using how "dangerous" the route is.
failed.

@Rakete1111 : this is another example which I would wish to be "solved" if the jump costs could become variable. If you really want to address the various issues, you might think about a heavier code change than what you did so far in your pull #2416. Dont get me wrong - and I am not sure - but eventually there might be a cleaner way.

**Comment by MessyMix** [comment]
I can't really tell—it's unclear from your post what you're trying to achieve. I played the save, looks like you are plotting a circular route; your second diagram is also confusing, because I can't relate that to the screenshot. I get it, the computer routing doesn't take into account the fuel of the Scramdrive, but how does the savegame demonstrate anything? What are you trying to do, and how does it show that the 150 fuel isn't included in the calculations?

tl;dr: i'm confused.

**Comment by EndrosG** [comment]
If you let the computer find the cheapest route from Polaris to Almach, it will always offer you the 3 jumps via links if you have either a jumpdrive or a scram drive installed. Only if you have neither of them but a jump drive only, then it will offer you the 2 jumps through korath territory.

The circle path in the savegame was just used to outline both routes in one picture. Yellow is what the game chose, red is what would be cheaper in this situation.

The selfmade pic is the result of my analysis how a situation needs to look like for this to be fully true. After I knew what to look for, I found the Polaris Almach route in the second attempt.
I used colors to separate jump paths from regular links.

You need in the minimum example a 3-link-path between source and destination and at least one of the waypoints on the linked path needs to be outside the neighborhood of source and destination, respectively. -- Schedar is out of range from Polaris and Achernar is in the neighborhood of Almach.
And you need a 2-jump-path with no direct links to source or destination.

The route Zubeneschamali - Aldhibain fails the first requirement. Both waypoints, Zubenelgenubi and Sabik, are neighbors of Zubeneschamali, thus the cheapest route with scram drive their would be a jump from Zubeneschamali zu Sabik and then a scram drive trip to Aldhibain. And thus the 2-jump-path via Unukalhai would not be cheaper.

There is another, 2nd example on the map, which I forgot about and just saw now again.
Izar - Spica. The game will always offer the 3-link-path via Acrux and Minka (if you have a hyperdrive or scram drive) but if you only have a jump drive, you get the 2-jump-path via Kraz.

**Comment by Elyssaen** [comment]
Yeah, I remember noticing the Scram Drive isn't perfectly supported in those calculations. We could probably do something about it reasonably easily. It's not _too_ noticeable with the three drives in the game, but it would be much more obvious in some content creation scenarios.

**Comment by endless-sky** [comment]
This would be relatively easy to do, but I wonder if more players would prefer for pathfinding to choose the quickest path (i.e. fewest total jumps), or to choose the path with the least amount of fuel expended?

**Comment by tehhowch** [comment]
Preference toggle :)

I would opt for fewest jumps as that means fewer salaries paid out, and waiting for escorts to arrive in-system is "free".

**Comment by EndrosG** [comment]
I would opt for fewest jumps too. Time is money. And we have deadlines. ;-)

In the example given above, *both* comes together:
the jump drive based path (2 jumps x 200 = 400 fuel) is faster and cheaper than the 3 jumps with scram drive (3x 150 = 450).

**Comment by endless-sky** [comment]
Fixed in 744cb0a, although right now I've got it set up to minimize fuel use rather than minimize travel time. (But, that would be easy enough to change if a large enough majority of people want it the other way.)

**Comment by jafdy** [comment]
Minimizing travel time is a relatively simple problem most of the time as the only time the path could be different is if you have both a jump drive and a hyperdrive.
